.github/workflows/e2e.yaml

```diff
@@ -1,59 +1,52 @@
 on:
   workflow_call:
     inputs:
-      k8sVersion:
-        description: version of k3s to use or "kind" and "kindIpv6"
-        type: string
-        required: true
-      target:
-        description: makefile target without e2e prefix
-        type: string
-        required: true
-      arch:
-        description: The golang arch
-        type: string
-        required: true
-      legacyKDS:
-        description: if should run tests with new implementation of KDS
-        type: string
+      matrix:
         required: true
-      cniNetworkPlugin:
-        description: The CNI networking plugin to use [flannel | calico]
         type: string
-        required: true
-      parallelism:
-        description: level of parallelization
+      runnersByArch:
         type: string
-        required: true
+        required: false
+        default: '{"amd64": "ubuntu-latest", "arm64": "circleci"}'
     secrets:
       circleCIToken:
         required: true
+env:
+  E2E_PARAM_K8S_VERSION: ${{ fromJSON(inputs.matrix).k8sVersion }}
+  E2E_PARAM_CNI_NETWORK_PLUGIN: ${{ fromJSON(inputs.matrix).cniNetworkPlugin }}
+  E2E_PARAM_ARCH: ${{ fromJSON(inputs.matrix).arch }}
+  E2E_PARAM_LEGACY_KDS: ${{ fromJSON(inputs.matrix).legacyKDS }}
+  E2E_PARAM_TARGET: ${{ fromJSON(inputs.matrix).target }}
+  E2E_PARAM_PARALLELISM: ${{ fromJSON(inputs.matrix).parallelism }}
 jobs:
   e2e:
-    runs-on: ubuntu-latest
+    # use the runner from the map, if the runner is circleci or '' then use ubuntu-latest
+    runs-on: ${{ contains(fromJSON('["circleci", ""]'), fromJSON(inputs.runnersByArch)[fromJSON(inputs.matrix).arch]) && 'ubuntu-latest' || fromJSON(inputs.runnersByArch)[fromJSON(inputs.matrix).arch]}}
```

Comment:
This expression is very smart!

Reply:
Yes it's a little too complicated but it's tricky to do this without having multiple jobs and if we do this the matrix becomes quickly unreadable